Implemented escaping newlines.

golddranks opened the pull request:
Implementation of feature suggestion mentioned in https://github.com/rust-lang/regex/issues/107 Tests and documentation included.

I think this is a nice, ergonomic bonus; the semantics are obvious and lightweight (and similar with the Rust strings newline escape). @BurntSushi said he doesn't have any strong feelings either way. What do others think?

It allows a case that was previously illegal, so not a breaking change.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @huonw (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

golddranks commented:
Rebased on newest master and added the compatibility for

> Rust's existing line continuation drops the \n and all proceeding whitespace up to the first non-whitespace character.

BurntSushi commented on regex-syntax/src/parser.rs:
```diff
@@ -185,6 +185,16 @@ impl Parser {
+                        while self.cur().is_whitespace() { self.bump(); };
+                        Ok(Build::Expr(Expr::Empty))
+                    }
+                    else { Err(self.err(ErrorKind::UnrecognizedEscape(c))) } },
```
Could you please format this code so that it's consistent with the rest of the code in this library? e.g., Consistent indentation, new lines after braces, etc.

golddranks commented:
Oops, I forgot the existence of test_err! and did some things with the expect_panic, and used .expect() to elaborate the message, but that breaks Rust 1.3 compatibility. Reverted already the tests to use test_err!, but going to revert the .expect() too now.

golddranks commented:
Or... do we want to support Rust 1.3? I noticed that `.expect()` is used elsewhere too, just not in the tests.

BurntSushi commented:
@golddranks Please don't increase the minimum required Rust version in this PR. `expect` exists as a method on `Option<T>` but not on `Result<T, E>` (pre Rust 1.4).

golddranks commented:
Ah, indeed, the already existing `.expect()`s were on `Option<T>`. Pardon me! I'll revert it.

BurntSushi commented:
Sorry for letting this PR languish, but I don't think I ever got to a place where I was comfortable with it. I don't think it's a good idea, in general, to introduce a syntactic element that *silently* behaves differently than it does in most other regex engines. I understand there are benefits to `\<newline>`, but I feel like the `x` flag covers *most* of them.

I know it's a small issue, and I think reasonable people can disagree. But at the end of the day, I feel a bit more comfortable with the conservative choice.
